modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/RunningStats.java

```diff
@@ -0,0 +1,311 @@
+ *
+ * http://prod.sandia.gov/techlib/access-control.cgi/2008/086212.pdf
+ *
+ * @internal
+ */
+public class RunningStats implements Writeable, Cloneable {
+    /** count of observations (same number of observations per field) */
+    protected long docCount = 0;
+    /** per field sum of observations */
+    protected HashMap<String, Double> fieldSum;
+    /** counts */
+    protected HashMap<String, Long> counts;
+    /** mean values (first moment) */
+    protected HashMap<String, Double> means;
+    /** variance values (second moment) */
+    protected HashMap<String, Double> variances;
+    /** skewness values (third moment) */
+    protected HashMap<String, Double> skewness;
+    /** kurtosis values (fourth moment) */
+    protected HashMap<String, Double> kurtosis;
+    /** covariance values */
+    protected HashMap<String, HashMap<String, Double>> covariances;
+
+    private RunningStats() {
+        init();
+    }
+
+    public RunningStats(final ArrayList<String> fieldNames, final double[] fieldVals) {
+        if (fieldVals != null && fieldVals.length > 0) {
+            init();
+            this.add(fieldNames, fieldVals);
+        }
+    }
+
+    private void init() {
+        counts = new HashMap<>();
+        fieldSum = new HashMap<>();
+        means = new HashMap<>();
+        skewness = new HashMap<>();
+        kurtosis = new HashMap<>();
+        covariances = new HashMap<>();
+        variances = new HashMap<>();
+    }
+
+    /** Ctor to create an instance of running statistics */
+    @SuppressWarnings("unchecked")
+    public RunningStats(StreamInput in) throws IOException {
+        this();
+        // read fieldSum
+        fieldSum = (HashMap<String, Double>)in.readGenericValue();
+        // counts
+        counts = (HashMap<String, Long>)in.readGenericValue();
+        // means
+        means = (HashMap<String, Double>)in.readGenericValue();
+        // variances
+        variances = (HashMap<String, Double>)in.readGenericValue();
+        // skewness
+        skewness = (HashMap<String, Double>)in.readGenericValue();
+        // kurtosis
+        kurtosis = (HashMap<String, Double>)in.readGenericValue();
+        // read covariances
+        covariances = (HashMap<String, HashMap<String, Double>>)in.readGenericValue();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        // marshall fieldSum
+        out.writeGenericValue(fieldSum);
+        // counts
+        out.writeGenericValue(counts);
+        // mean
+        out.writeGenericValue(means);
+        // variances
+        out.writeGenericValue(variances);
+        // skewness
+        out.writeGenericValue(skewness);
+        // kurtosis
+        out.writeGenericValue(kurtosis);
+        // covariances
+        out.writeGenericValue(covariances);
+    }
+
+    /** create an empty instance */
+    protected static RunningStats EMPTY() {
+        return new RunningStats();
+    }
```

Comment:
when I see this method, it makes me think of a singleton that represents empty stats and should not be modified. Maybe we should remove this method and just expect callers to call the constructor?